pkg/threescale/threescale.go

```diff
@@ -115,68 +117,95 @@ func (s *Threescale) extractProxyConf(cfg *config.Params, c *sysC.ThreeScaleClie
 	return pce, proxyConfErr
 }
 
-// isAuthorized returns code 0 if authorization is successful
-// based on grpc return codes https://github.com/grpc/grpc-go/blob/master/codes/codes.go
-func (s *Threescale) isAuthorized(cfg *config.Params, request authorization.InstanceMsg, proxyConf sysC.ProxyConfigElement, client *backendC.ThreeScaleClient) (rpc.Status, error) {
+// isAuthorized - is responsible for parsing the incoming request and determining if it is valid, building out the request to be sent
+// to 3scale and parsing the response. Returns code 0 if authorization is successful based on
+// grpc return codes https://github.com/grpc/grpc-go/blob/master/codes/codes.go
+func (s *Threescale) isAuthorized(svcID string, request authorization.InstanceMsg, proxyConf sysC.ProxyConfig, client *backendC.ThreeScaleClient) rpc.Status {
+	var (
+		// Application ID authentication pattern - App Key is optional when using this authn
+		appID, appKey string
+
+		// Application Key auth pattern
+		userKey string
+
+		// Function to be called when authn patter has been determined
+		authRep authRepFn
+	)
+
 	if request.Action.Path == "" {
-		return status.WithInvalidArgument("missing request path"), nil
+		request.Action.Path = "/"
 	}
 
-	userKey, err := parseActionPath(request.Action.Path)
-	if err != nil {
-		return status.WithPermissionDenied(err.Error()), nil
+	metrics := generateMetrics(request.Action.Path, request.Action.Method, proxyConf)
+	if len(metrics) == 0 {
+		return status.WithPermissionDenied(fmt.Sprintf("no matching mapping rule for request with method %s and path %s",
+			request.Action.Method, request.Action.Path))
+	}
+
+	authRepRequest := authRepRequest{
+		svcID: svcID,
+		auth: backendC.TokenAuth{
+			Type:  proxyConf.Content.BackendAuthenticationType,
+			Value: proxyConf.Content.BackendAuthenticationValue,
+		}}
+
+	if request.Subject != nil {
```

Comment:
if `Subject == nil` we can probably error out earlier, before matching rules, right?

Reply:
True, Again this is just covering a messed up config and avoiding a panic but yes, I'll move this to fail early

Reply:
Fixed